beacon-chain/p2p/gossipsub_peer_controller.go

```diff
@@ -0,0 +1,153 @@
+package p2p
+
+import (
+	"context"
+	"math"
+	"sync"
+	"time"
+
+	"github.com/OffchainLabs/prysm/v7/beacon-chain/p2p/gossipsubcrawler"
+	"github.com/OffchainLabs/prysm/v7/cmd/beacon-chain/flags"
+	"github.com/ethereum/go-ethereum/p2p/enode"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/pkg/errors"
+)
+
+const (
+	peerPerTopic = 20
+)
+
+type GossipsubPeerDialer struct {
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	listPeersFunc func(topic string) []peer.ID
+	dialPeersFunc func(ctx context.Context, maxConcurrentDials int, nodes []*enode.Node) uint
+
+	crawler        gossipsubcrawler.Crawler
+	topicsProvider gossipsubcrawler.SubnetTopicsProvider
+
+	wg   sync.WaitGroup
+	once sync.Once
+}
+
+func NewGossipsubPeerDialer(
+	crawler gossipsubcrawler.Crawler,
+	listPeers func(topic string) []peer.ID,
+	dialPeers func(ctx context.Context, maxConcurrentDials int, nodes []*enode.Node) uint,
+) *GossipsubPeerDialer {
+	ctx, cancel := context.WithCancel(context.Background())
+	return &GossipsubPeerDialer{
+		listPeersFunc: listPeers,
+		dialPeersFunc: dialPeers,
+		crawler:       crawler,
+		ctx:           ctx,
+		cancel:        cancel,
+	}
+}
+
+func (g *GossipsubPeerDialer) Stop() {
+	g.cancel()
+	g.wg.Wait()
+}
+
+func (g *GossipsubPeerDialer) Start(provider gossipsubcrawler.SubnetTopicsProvider) error {
+	if provider == nil {
+		return errors.New("topics provider is nil")
+	}
```

Comment:
Safety is good... but I find it a little odd we have to check for nil here. Why would this be called with a nil provider arg?